relation_engine_server/utils/ensure_specs.py

```diff
@@ -0,0 +1,265 @@
+        "views": failed_views_names,
+        "analyzers": failed_analyzers_names,
+    }
+
+
+def get_names(specs, schema_type):
+    """
+    Given views/analyzers/collections, collate names using required properties
+    """
+    names = []
+    if schema_type in ["views", "analyzers"]:
+        for spec in specs:
+            names.append(f"{spec['name']}/{spec['type']}")
+    elif schema_type in ["indexes"]:
+        for coll_name, indexes in specs.items():
+            for index in indexes:
+                names.append(f"{coll_name}/{index['type']}/{index['fields']}")
+    else:
+        raise RuntimeError(f'Unknown schema type "{schema_type}"')
+    return names
+
+
+def print_failed_vs_server(schema_type, failed_specs, server_specs):
+    """
+    Print message with names and contents of failed local specs and all server specs
+    """
+    dec = "*" * 80
+
+    def format_json(jo):
+        return json.dumps(jo, indent=4)
+
+    fail_msg = (
+        dec + "\n"
+        f"----------> failed ({len(failed_specs)} {schema_type}) ---------->"
+        "\n"
+        f"----------> names: {get_names(failed_specs, schema_type)} ---------->"
+        "\n" + format_json(failed_specs) + "\n"
+        f"----------> server ({len(server_specs)} {schema_type}) ---------->"
+        "\n"
+        f"----------> names: {get_names(server_specs, schema_type)} ---------->"
+        "\n" + format_json(server_specs) + "\n" + dec
+    )
+
+    print(fail_msg)
+
+
+def round_float(num: float) -> float:
+    """
+    For round-off error in floats
+    Arbitrarily chose 7 places
+    """
+    return round(num, 7)
+
+
+def excise_namespace(analyzer_name: str) -> str:
+    """
+    Remove namespace prefix, e.g.,
+    namespace::thing -> thing
+    """
+    return analyzer_name.split("::")[-1]
+
+
+def mod_obj_literal(
+    spec_unit: Union[list, dict],
+    literal_type: type,
+    func: Callable[[Union[float, str]], Union[float, str]],
+) -> None:
+    """
+    Modify dict in-place recursively
+    Some specs won't match because of
+    * round-off error in floats
+    * namespacing in analyzers, e.g., "_system::icu_tokenize"
+
+    Parameters
+    ----------
+    spec_unit -     recursively accessed data structure unit of JSON obj
+    literal_type -  str or float
+    func -          function called to modify that str or float in-place
+    """
+    if isinstance(spec_unit, dict):
+        for k, v in spec_unit.items():
+            if isinstance(v, dict) or isinstance(v, list):
+                mod_obj_literal(v, literal_type, func)
+            elif isinstance(v, literal_type):
+                spec_unit[k] = func(v)  # type: ignore
+    elif isinstance(spec_unit, list):
+        for i, v in enumerate(spec_unit):
+            if isinstance(v, dict) or isinstance(v, list):
+                mod_obj_literal(v, literal_type, func)
+            elif isinstance(v, literal_type):
+                spec_unit[i] = func(v)  # type: ignore
```

Comment:
hmmm, this looks familiar...

Reply:
My reasoning is premature optimization. And it's a small sin and may increase readability, even if it's an eyesore ...